### PR TITLE
[MA-27] Movie Genres Operations

### PR DESCRIPTION
**Pull request by cankurtur**
#34

### CODE REVIEW

File: movieApp/movieApp/Network/APIConstants.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+//  APIConstants.swift
+//  movieApp
+//
+//  Created by Can Kurtur on 3.09.2021.
+//
+
+import Foundation
+
+struct APIConstants {
+    static let baseURL = "https://api.themoviedb.org/3/movie/popular?"
```
Comment (batuhanabay) [inline movieApp/movieApp/Network/APIConstants.swift]: baseURL should not be like that. This one is base of popular movies url, isn't it?

File: movieApp/movieApp/Network/ResponseModels/MovieGenresResponseModel.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//  Created by Can Kurtur on 5.09.2021.
+//
+
+import Foundation
+
+struct MovieGenresResponseModel: Codable {
+    let genres: [GenresResponseModel]
+}
+
+struct GenresResponseModel: Codable {
```
Comment (batuhanabay) [inline movieApp/movieApp/Network/ResponseModels/MovieGenresResponseModel.swift]: This is a single genre model so its name should be singular.

File: movieApp/movieApp/Scenes/Main/MainViewController.swift [hunk 3]
```diff
@@ -23,38 +23,50 @@ class MainViewController: UIViewController {
...
-        scrollView.delegate = self
-        viewModel = MainViewModel()
-        filmTableView.register(UINib(nibName: Constants.cellNibName, bundle: nil), forCellReuseIdentifier: Constants.cellIdentifier)
-        tableViewHeightConstraint.constant = CGFloat(count * Constants.tableViewRowHeight)
         setupUI()
+        setupBinding()
     }
 
     // MARK: - Business Logic
     func setupUI() {
```
Comment (batuhanabay) [inline movieApp/movieApp/Scenes/Main/MainViewController.swift]: this method should be private

File: movieApp/movieApp/Scenes/Main/MainViewController.swift [hunk 4]
```diff
@@ -23,38 +23,50 @@ class MainViewController: UIViewController {
...
+        scrollView.delegate = self
         view.backgroundColor = .whiteTwo
     }
+
+    func setupBinding() {
```
Comment (batuhanabay) [inline movieApp/movieApp/Scenes/Main/MainViewController.swift]: this method should be private

File: movieApp/movieApp/Scenes/Main/MainViewController.swift [hunk 5]
```diff
@@ -23,38 +23,50 @@ class MainViewController: UIViewController {
...
+        viewModel.getPopularMovies { cardViewUIModel in
+            self.cardViewUIModel = cardViewUIModel
+            self.count = Double(cardViewUIModel.count)
+            self.filmTableView.reloadData()
+            self.setupTableView()
+        }
+    }
+
+    func setupTableView() {
```
Comment (batuhanabay) [inline movieApp/movieApp/Scenes/Main/MainViewController.swift]: this method should be private

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift [hunk 6]
```diff
@@ -9,20 +9,113 @@ import UIKit
...
+            }
+        }
+    }
+
+    func genresIdToString(intArray: [Int]) -> String {
+        var detailsString = ""
+        var editedString = ""
+        if genreUserDefault.checkUserDefault(key: UserDefaultsConstant.genresDict) {
+            let dict = genreUserDefault.getUserDefault(key: UserDefaultsConstant.genresDict)
+            for index in 0..<(intArray.count) {
```
Comment (batuhanabay) [inline movieApp/movieApp/Scenes/Main/MainViewModel.swift]: I remember that @gidercagri  said "you should check and learn for-each loop and change this old index-based  for loop to for-each loop" so please check it and change this loop

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift [hunk 7]
```diff
@@ -9,20 +9,113 @@ import UIKit
...
+                    detailsString += "\(value), "
+                }
+            }
+            editedString = String(detailsString.dropLast(2))
+        }
+        return editedString
+    }
+}
+
+struct UserDefaultsConstant {
```
Comment (batuhanabay) [inline movieApp/movieApp/Scenes/Main/MainViewModel.swift]: We should move userDefaults related class and structs into separate file

**Review by batuhanabay — COMMENTED** [review]
Added comments should be resolved

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift [hunk 8]
```diff
@@ -9,20 +9,113 @@ import UIKit
...
+
+    func getUserDefault(key: String) -> [String: String] {
+        if let object = userDefaults.object(forKey: key) {
+            return object as? [String: String] ?? ["": ""]
+        } else {
+            return ["": ""]
+        }
+    }
+
+    func checkUserDefault(key: String) -> Bool {
```
Comment (gidercagri) [inline movieApp/movieApp/Scenes/Main/MainViewModel.swift]: checkUserDefault already getting data from UserDefaults, do we really need to use getUserDefault? Maybe we can directly try to get data and if we can't then continue with [:]

File: movieApp/movieApp/Network/APIConstants.swift (hunk 1 above)
Comment (cankurtur) [reply movieApp/movieApp/Network/APIConstants.swift]: You right. I modified this struct.

File: movieApp/movieApp/Network/ResponseModels/MovieGenresResponseModel.swift (hunk 2 above)
Comment (cankurtur) [reply movieApp/movieApp/Network/ResponseModels/MovieGenresResponseModel.swift]: Fixed.

File: movieApp/movieApp/Scenes/Main/MainViewController.swift (hunk 3 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewController.swift]: Fixed.

File: movieApp/movieApp/Scenes/Main/MainViewController.swift (hunk 4 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewController.swift]: Fixed.

File: movieApp/movieApp/Scenes/Main/MainViewController.swift (hunk 5 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewController.swift]: Fixed.

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift (hunk 7 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewModel.swift]: Fixed.

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift (hunk 6 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewModel.swift]: fixed.

**Review by batuhanabay — APPROVED** [review]

**Review by gidercagri — APPROVED** [review]

File: movieApp/movieApp/Scenes/Main/MainViewModel.swift (hunk 8 above)
Comment (cankurtur) [reply movieApp/movieApp/Scenes/Main/MainViewModel.swift]: Sonraki PR'da bakilacak

**Pull request merged**